diff -Nur -x '*.orig' -x '*~' postgresql-8.5-8.5~alpha1/GNUmakefile.in postgresql-8.5-8.5~alpha1.new/GNUmakefile.in
--- postgresql-8.5-8.5~alpha1/GNUmakefile.in	2009-08-15 00:37:18.000000000 +0300
+++ postgresql-8.5-8.5~alpha1.new/GNUmakefile.in	2009-09-03 23:25:58.000000000 +0300
@@ -12,6 +12,7 @@
 	$(MAKE) -C doc all
 	$(MAKE) -C src all
 	$(MAKE) -C config all
+	$(MAKE) -C contrib all
 	@echo "All of PostgreSQL successfully made. Ready to install."
 
 html man:
@@ -21,12 +22,14 @@
 	$(MAKE) -C doc $@
 	$(MAKE) -C src $@
 	$(MAKE) -C config $@
+	$(MAKE) -C contrib $@
 	@echo "PostgreSQL installation complete."
 
 installdirs uninstall coverage:
 	$(MAKE) -C doc $@
 	$(MAKE) -C src $@
 	$(MAKE) -C config $@
+	$(MAKE) -C contrib $@
 
 distprep:
 	$(MAKE) -C doc $@
