diff -Nur postgresql-8.3/build-tree/postgresql-8.3beta1/src/bin/psql/startup.c postgresql-8.3.new/build-tree/postgresql-8.3beta1/src/bin/psql/startup.c
--- postgresql-8.3beta1/src/bin/psql/startup.c	2007-10-09 15:34:58.000000000 +0200
+++ postgresql-8.3beta1/src/bin/psql/startup.c	2007-10-09 15:36:17.000000000 +0200
@@ -212,7 +212,7 @@
 
 		if (PQstatus(pset.db) == CONNECTION_BAD &&
 			PQconnectionUsedPassword(pset.db) &&
-			password == NULL &&
+			password == NULL && !getenv("PGPASSWORD") &&
 			!feof(stdin))
 		{
 			PQfinish(pset.db);
