diff -Nur postgresql-8.2/build-tree/postgresql-8.2.3/src/timezone/Makefile postgresql-8.2.new/build-tree/postgresql-8.2.3/src/timezone/Makefile
--- postgresql-8.2.3/src/timezone/Makefile	2006-07-25 05:51:22.000000000 +0200
+++ postgresql-8.2.3/src/timezone/Makefile	2007-03-30 18:52:54.000000000 +0200
@@ -34,7 +34,7 @@
 	$(CC) $(CFLAGS) $(ZICOBJS) $(LDFLAGS) $(LIBS) -o $@$(X)
 
 install: all installdirs
-	./zic -d '$(DESTDIR)$(datadir)/timezone' $(TZDATAFILES)
+	ln -s /usr/share/zoneinfo '$(DESTDIR)$(datadir)/timezone'
 	$(MAKE) -C tznames $@
 
 installdirs:
