diff -Nur -x '*.orig' -x '*~' postgresql-8.5-8.5~alpha2/src/test/regress/pg_regress.c postgresql-8.5-8.5~alpha2.new/src/test/regress/pg_regress.c
--- postgresql-8.5-8.5~alpha2/src/test/regress/pg_regress.c	2009-08-18 13:30:41.000000000 +0300
+++ postgresql-8.5-8.5~alpha2.new/src/test/regress/pg_regress.c	2009-11-05 23:52:05.000000000 +0200
@@ -889,7 +889,7 @@
 
 	/* And now we can build and execute the shell command */
 	snprintf(psql_cmd, sizeof(psql_cmd),
-			 SYSTEMQUOTE "\"%s%spsql\" -X -c \"%s\" \"%s\"" SYSTEMQUOTE,
+			 SYSTEMQUOTE "\"%s%spsql\" -h /tmp -X -c \"%s\" \"%s\"" SYSTEMQUOTE,
 			 psqldir ? psqldir : "",
 			 psqldir ? "/" : "",
 			 query_escaped,
@@ -2139,7 +2139,7 @@
 		 * Check if there is a postmaster running already.
 		 */
 		snprintf(buf2, sizeof(buf2),
-				 SYSTEMQUOTE "\"%s/psql\" -X postgres <%s 2>%s" SYSTEMQUOTE,
+				 SYSTEMQUOTE "\"%s/psql\" -h /tmp -X postgres <%s 2>%s" SYSTEMQUOTE,
 				 bindir, DEVNULL, DEVNULL);
 
 		for (i = 0; i < 16; i++)
@@ -2171,7 +2171,7 @@
 		 */
 		header(_("starting postmaster"));
 		snprintf(buf, sizeof(buf),
-				 SYSTEMQUOTE "\"%s/postgres\" -D \"%s/data\" -F%s -c \"listen_addresses=%s\" > \"%s/log/postmaster.log\" 2>&1" SYSTEMQUOTE,
+				 SYSTEMQUOTE "\"%s/postgres\" -D \"%s/data\" -k /tmp -F%s -c \"listen_addresses=%s\" > \"%s/log/postmaster.log\" 2>&1" SYSTEMQUOTE,
 				 bindir, temp_install,
 				 debug ? " -d 5" : "",
 				 hostname ? hostname : "",
diff -Nur -x '*.orig' -x '*~' postgresql-8.5-8.5~alpha2/src/test/regress/pg_regress_main.c postgresql-8.5-8.5~alpha2.new/src/test/regress/pg_regress_main.c
--- postgresql-8.5-8.5~alpha2/src/test/regress/pg_regress_main.c	2009-10-14 00:04:01.000000000 +0300
+++ postgresql-8.5-8.5~alpha2.new/src/test/regress/pg_regress_main.c	2009-11-05 23:52:58.000000000 +0200
@@ -59,7 +59,7 @@
 	add_stringlist_item(expectfiles, expectfile);
 
 	snprintf(psql_cmd, sizeof(psql_cmd),
-			 SYSTEMQUOTE "\"%s%spsql\" -X -a -q -P linestyle=ascii -d \"%s\" < \"%s\" > \"%s\" 2>&1" SYSTEMQUOTE,
+			 SYSTEMQUOTE "\"%s%spsql\" -h /tmp -X -a -q -P linestyle=ascii -d \"%s\" < \"%s\" > \"%s\" 2>&1" SYSTEMQUOTE,
 			 psqldir ? psqldir : "",
 			 psqldir ? "/" : "",
 			 dblist->str,
