diff -Nur postgresql-8.2/build-tree/postgresql-8.2.4/configure postgresql-8.2.new/build-tree/postgresql-8.2.4/configure
--- postgresql-8.2.4/configure	2007-04-20 05:20:41.000000000 +0200
+++ postgresql-8.2.4/configure	2007-06-20 18:22:04.000000000 +0200
@@ -7537,7 +7537,7 @@
 
 
 
-for ac_header in crypt.h dld.h endian.h fp_class.h getopt.h ieeefp.h langinfo.h poll.h pwd.h sys/ipc.h sys/poll.h sys/pstat.h sys/resource.h sys/select.h sys/sem.h sys/socket.h sys/shm.h sys/time.h sys/un.h termios.h utime.h wchar.h wctype.h kernel/OS.h kernel/image.h SupportDefs.h
+for ac_header in crypt.h dld.h endian.h fp_class.h getopt.h ieeefp.h langinfo.h poll.h pwd.h sys/ipc.h sys/poll.h sys/pstat.h sys/resource.h sys/select.h sys/sem.h sys/socket.h sys/shm.h sys/tas.h sys/time.h sys/un.h termios.h utime.h wchar.h wctype.h kernel/OS.h kernel/image.h SupportDefs.h
 do
 as_ac_Header=`echo "ac_cv_header_$ac_header" | $as_tr_sh`
 if eval "test \"\${$as_ac_Header+set}\" = set"; then
diff -Nur postgresql-8.2/build-tree/postgresql-8.2.4/src/include/pg_config.h.in postgresql-8.2.new/build-tree/postgresql-8.2.4/src/include/pg_config.h.in
--- postgresql-8.2.4/src/include/pg_config.h.in	2006-11-06 04:44:38.000000000 +0100
+++ postgresql-8.2.4/src/include/pg_config.h.in	2007-06-20 18:22:30.000000000 +0200
@@ -484,6 +484,9 @@
 /* Define to 1 if you have the <sys/stat.h> header file. */
 #undef HAVE_SYS_STAT_H
 
+/* Define to 1 if you have the <sys/tas.h> header file. */
+#undef HAVE_SYS_TAS_H
+
 /* Define to 1 if you have the <sys/time.h> header file. */
 #undef HAVE_SYS_TIME_H
 
diff -Nur postgresql-8.2/build-tree/postgresql-8.2.4/src/include/storage/s_lock.h postgresql-8.2.new/build-tree/postgresql-8.2.4/src/include/storage/s_lock.h
--- postgresql-8.2.4/src/include/storage/s_lock.h	2006-06-08 00:24:45.000000000 +0200
+++ postgresql-8.2.4/src/include/storage/s_lock.h	2007-06-20 18:23:04.000000000 +0200
@@ -556,6 +556,18 @@
 #endif /* __mips__ && !__sgi */
 
 
+#if defined(__m32r__) && defined(HAVE_SYS_TAS_H)	/* Renesas' M32R */
+#define HAS_TEST_AND_SET
+
+#include <sys/tas.h>
+
+typedef int slock_t;
+
+#define TAS(lock) tas(lock)
+
+#endif /* __m32r__ */
+
+
 /* These live in s_lock.c, but only for gcc */
 
 
