--- a/src/test/regress/pg_regress.c
+++ b/src/test/regress/pg_regress.c
@@ -828,6 +828,7 @@ initialize_environment(void)
 	{
 		const char *pghost;
 		const char *pgport;
+		char       pgportstr[32];
 
 		/*
 		 * When testing an existing install, we honor existing environment
@@ -858,14 +859,18 @@ initialize_environment(void)
 			pghost = "localhost";
 #endif
 
-		if (pghost && pgport)
-			printf(_("(using postmaster on %s, port %s)\n"), pghost, pgport);
-		if (pghost && !pgport)
-			printf(_("(using postmaster on %s, default port)\n"), pghost);
-		if (!pghost && pgport)
-			printf(_("(using postmaster on Unix socket, port %s)\n"), pgport);
-		if (!pghost && !pgport)
-			printf(_("(using postmaster on Unix socket, default port)\n"));
+		/* Precompute the "port xxx" part so we don't have 6 printf()s below */
+		if (pgport)
+			snprintf(pgportstr, sizeof(pgportstr), _("port %s"), pgport);
+		else
+			snprintf(pgportstr, sizeof(pgportstr), _("default port"));
+
+		if (pghost && *pghost != '/')
+			printf(_("(using postmaster on %s, %s)\n"), pghost, pgportstr);
+		else if (pghost)
+			printf(_("(using postmaster on Unix socket %s, %s)\n"), pghost, pgportstr);
+		else
+			printf(_("(using postmaster on Unix socket, %s)\n"), pgportstr);
 	}
 
 	convert_sourcefiles();
@@ -2238,7 +2243,9 @@ regression_main(int argc, char *argv[],
 		 */
 		header(_("starting postmaster"));
 		snprintf(buf, sizeof(buf),
-				 "\"%s/postgres\" -D \"%s/data\" -F%s -c \"listen_addresses=%s\" > \"%s/log/postmaster.log\" 2>&1",
+				 hostname && *hostname == '/'
+					 ? "\"%s/postgres\" -D \"%s/data\" -F%s -k \"%s\" > \"%s/log/postmaster.log\" 2>&1"
+					 : "\"%s/postgres\" -D \"%s/data\" -F%s -c \"listen_addresses=%s\" > \"%s/log/postmaster.log\" 2>&1",
 				 bindir, temp_install,
 				 debug ? " -d 5" : "",
 				 hostname ? hostname : "",
