Index: postgresql-9.2/src/backend/main/main.c
===================================================================
--- postgresql-9.2.orig/src/backend/main/main.c	2012-05-15 07:18:26.105536454 +0200
+++ postgresql-9.2/src/backend/main/main.c	2012-05-15 07:41:06.401602291 +0200
@@ -335,7 +335,7 @@
 check_root(const char *progname)
 {
 #ifndef WIN32
-	if (geteuid() == 0)
+	if (0 && geteuid() == 0)
 	{
 		write_stderr("\"root\" execution of the PostgreSQL server is not permitted.\n"
 					 "The server must be started under an unprivileged user ID to prevent\n"
Index: postgresql-9.2/src/bin/initdb/initdb.c
===================================================================
--- postgresql-9.2.orig/src/bin/initdb/initdb.c	2012-05-15 07:18:26.105536454 +0200
+++ postgresql-9.2/src/bin/initdb/initdb.c	2012-05-15 07:41:06.401602291 +0200
@@ -581,7 +581,7 @@
 
 	struct passwd *pw;
 
-	if (geteuid() == 0)			/* 0 is root's uid */
+	if (0 && geteuid() == 0)		/* 0 is root's uid */
 	{
 		fprintf(stderr,
 				_("%s: cannot be run as root\n"
Index: postgresql-9.2/src/bin/pg_ctl/pg_ctl.c
===================================================================
--- postgresql-9.2.orig/src/bin/pg_ctl/pg_ctl.c	2012-05-15 07:18:26.109536453 +0200
+++ postgresql-9.2/src/bin/pg_ctl/pg_ctl.c	2012-05-15 07:41:06.401602291 +0200
@@ -1997,7 +1997,7 @@
 	 * Disallow running as root, to forestall any possible security holes.
 	 */
 #ifndef WIN32
-	if (geteuid() == 0)
+	if (0 && geteuid() == 0)
 	{
 		write_stderr(_("%s: cannot be run as root\n"
 					   "Please log in (using, e.g., \"su\") as the "
